source/ProcessManager.Orchestrations/Processes/BRS_021/Shared/ElectricityMarket/MeteringPointMasterDataProvider.cs

```diff
@@ -75,111 +75,81 @@ internal async Task<IReadOnlyCollection<MeteringPointMasterData>> GetMasterData(
+                        new ElectricityMarketModels.MeteringPointIdentification(parentId),
+                        new Interval(from, to))
+                    .ConfigureAwait(false))
+                .ToList();
 
-            // If no parent data was fetched, fall back to child supplier
-            if (!parentData.TryGetValue(parentId, out var parentEntries) || parentEntries.Count == 0)
+            // If no parent data was fetched, fall back to own energy supplier (even if null)
+            if (parentMasterData.Count == 0)
             {
+                // TODO: Should this throw an exception instead?
                 result.Add(
-                    CreatePmMeteringPointMasterData(
-                        child,
+                    CreateMeteringPointMasterData(
+                        meteringPointMasterData,
                         from.ToDateTimeOffset(),
                         to.ToDateTimeOffset(),
                         new MeteringPointId(parentId),
-                        child.EnergySupplier is not null ? ActorNumber.Create(child.EnergySupplier) : null));
+                        meteringPointMasterData.EnergySupplier is not null
+                            ? ActorNumber.Create(meteringPointMasterData.EnergySupplier)
+                            : null));
                 continue;
             }
 
-            // STEP 1: Identify time boundaries where supplier might change
-            var sliceBoundaries = new SortedSet<Instant> { from, to };
+            // STEP 1: Identify time boundaries where energy supplier might change
+            var parentMasterDataPeriods = new List<Interval>();
 
-            foreach (var parent in parentEntries)
+            foreach (var parent in parentMasterData.OrderBy(p => p.ValidFrom).ThenBy(p => p.ValidTo))
             {
-                var overlapsChild = parent.ValidFrom < to && parent.ValidTo > from;
-                if (!overlapsChild) continue;
-
                 var overlapStart = Instant.Max(from, parent.ValidFrom);
                 var overlapEnd = Instant.Min(to, parent.ValidTo);
 
-                sliceBoundaries.Add(overlapStart);
-                sliceBoundaries.Add(overlapEnd);
+                var period = new Interval(overlapStart, overlapEnd);
+
+                parentMasterDataPeriods.Add(period);
             }
 
             // STEP 2: Build one segment per interval between boundaries
-            var orderedBoundaries = sliceBoundaries.OrderBy(b => b).ToList();
-
-            for (var i = 0; i < orderedBoundaries.Count - 1; i++)
+            foreach (var currentPeriod in parentMasterDataPeriods)
             {
-                var sliceFrom = orderedBoundaries[i];
-                var sliceTo = orderedBoundaries[i + 1];
-
                 // Find parent supplier valid in this slice
-                var matchingParent = parentEntries.FirstOrDefault(
+                var matchingParent = parentMasterData.Single(
```

Comment:
```suggestion
                var matchingParent = parentMasterData.Single(
```

We could catch if this `Single()` throws an error, then log the masterdata which would have made the `Single()` fail.